### PR TITLE
Fix `instantiate` not installing artifacts

### PR DESCRIPTION
**Pull request by staticfloat**
Also contains the "verbose mode" downloading from a previous PR, now more fully tested.

### CODE REVIEW

**Comment by staticfloat** [comment]
bors try

File: src/API.jl [hunk 1]
```diff
@@ -724,7 +725,10 @@ function instantiate(ctx::Context; manifest::Union{Bool, Nothing}=nothing,
                  " Otherwise, remove `$name` with `Pkg.rm(\"$name\")`.",
                  " Finally, run `Pkg.instantiate()` again.")
     end
+    # Download artifacts for this here package before fast-exiting
+    Operations.download_artifacts(ctx, [dirname(ctx.env.manifest_file)]; platform=platform, verbose=verbose)
     Operations.is_instantiated(ctx) && return
```
Comment (fredrikekre) [inline src/API.jl]: `is_instantiated` probably needs to be updated too, right?

Comment (staticfloat) [reply src/API.jl]: I guess it depends on what you want it to do; would we want it to return `false` if any artifacts in any dependencies are not instantiated?  That kind of changes the meaning a bit, but would make sense to me.

Comment (fredrikekre) [reply src/API.jl]: Yea, otherwise we exit here without functioning packages.

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
bors try-

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
The Julia 1.3 run on OSX seems really, really flaky.  That's strange to me.

**Comment by fredrikekre** [comment]
Failure on AV looks legit.

**Comment by KristofferC** [comment]
Master doesnt seem to be fundamentally broken, for example https://github.com/JuliaLang/Pkg.jl/pull/1381 passed all tests.

**Comment by staticfloat** [comment]
I guess that's because the version of Julia I was using was incompatible with `master` perhaps?  The tests for `master` weren't working for me when I ran locally.  I'll rebase this on top of `master` again then.

**Comment by staticfloat** [comment]
Okay, I rebased on top of `master` and testing locally gets this:

```
 Resolving package versions...
  Updating `/tmp/jl_dtuh1f/Project.toml`
  [7876af07] + Example v0.5.3
  Updating `/tmp/jl_dtuh1f/Manifest.toml`
  [7876af07] + Example v0.5.3
 Installed Example ─ v0.2.0
pinning / freeing: Test Failed at /home/sabae/src/Jotunheimr/Pkg.jl/test/pkg.jl:184
  Expression: ((Pkg.dependencies())[TEST_PKG.uuid]).version.minor == 2
   Evaluated: 0x00000005 == 2
Stacktrace:
 [1] macro expansion at /home/sabae/src/Jotunheimr/Pkg.jl/test/pkg.jl:184 [inlined]
 [2] macro expansion at /home/sabae/tmp/julia-build/julia-master/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1107 [inlined]
 [3] (::Main.PkgTests.OperationsTest.var"##1#12")(::String) at /home/sabae/src/Jotunheimr/Pkg.jl/test/pkg.jl:181
  Updating registry at `/tmp/jl_snVNfl/registries/General`
  Updating git-repo `/home/sabae/src/Jotunheimr/Pkg.jl/test/registries/General`
  Updating `/tmp/jl_dtuh1f/Project.toml`
 [no changes]
  Updating `/tmp/jl_dtuh1f/Manifest.toml`
 [no changes]
pinning / freeing: Test Failed at /home/sabae/src/Jotunheimr/Pkg.jl/test/pkg.jl:186
  Expression: ((Pkg.dependencies())[TEST_PKG.uuid]).version.minor == 2
   Evaluated: 0x00000005 == 2
Stacktrace:
 [1] macro expansion at /home/sabae/src/Jotunheimr/Pkg.jl/test/pkg.jl:186 [inlined]
 [2] macro expansion at /home/sabae/tmp/julia-build/julia-master/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1107 [inlined]
 [3] (::Main.PkgTests.OperationsTest.var"##1#12")(::String) at /home/sabae/src/Jotunheimr/Pkg.jl/test/pkg.jl:181
  Updating `/tmp/jl_dtuh1f/Project.toml`
 [no changes]
  Updating `/tmp/jl_dtuh1f/Manifest.toml`
 [no changes]
  Updating registry at `/tmp/jl_snVNfl/registries/General`
  Updating git-repo `/home/sabae/src/Jotunheimr/Pkg.jl/test/registries/General`
  Updating `/tmp/jl_dtuh1f/Project.toml`
 [no changes]
  Updating `/tmp/jl_dtuh1f/Manifest.toml`
 [no changes]
  Updating `/tmp/jl_dtuh1f/Project.toml`
  [7876af07] - Example v0.5.3
  Updating `/tmp/jl_dtuh1f/Manifest.toml`
  [7876af07] - Example v0.5.3
Test Summary:     | Pass  Fail  Total
pinning / freeing |    1     2      3
ERROR: LoadError: LoadError: Some tests did not pass: 1 passed, 2 failed, 0 errored, 0 broken.
in expression starting at /home/sabae/src/Jotunheimr/Pkg.jl/test/pkg.jl:124
in expression starting at /home/sabae/src/Jotunheimr/Pkg.jl/test/runtests.jl:9
ERROR: Package Pkg errored during testing
```

**Comment by staticfloat** [comment]
Hmmm, I didn't think these changes could have been affecting that, but it appears that it on `master` that test does not fail.

**Comment by KristofferC** [comment]
That test error does seem a bit odd. It is pinning `Example` to 0.2, checking that version 0.2 is installed and then updating and checking that 0.2 is still installed.

**Comment by staticfloat** [comment]
I figured that one out; it was a bad rebase.  I have a different one now, but I am convinced that it's my fault.  :)

**Comment by staticfloat** [comment]
This fails locally for me with the following:

```
Pkg.add: Test Failed at /home/sabae/src/Jotunheimr/Pkg.jl/test/api.jl:248                                                                                                                                   
  Expression: ((Pkg.dependencies())[parsers_uuid]).version == original_parsers_version                                                                                                                      
   Evaluated: v"0.3.6" == v"0.3.7"                                                                                                                                                                          
Stacktrace:                                                                                                                                                                                                 
 [1] (::Main.PkgTests.APITests.var"##60#70")(::String) at /home/sabae/src/Jotunheimr/Pkg.jl/test/api.jl:248                                                                                                 
 [2] (::Main.PkgTests.Utils.var"##4#7"{Bool,Main.PkgTests.APITests.var"##60#70"})() at /home/sabae/src/Jotunheimr/Pkg.jl/test/utils.jl:42                                                                   
 [3] withenv(::Main.PkgTests.Utils.var"##4#7"{Bool,Main.PkgTests.APITests.var"##60#70"}, ::Pair{String,Nothing}, ::Vararg{Pair{String,Nothing},N} where N) at ./env.jl:161                                  
 [4] #temp_pkg_dir#1(::Bool, ::typeof(Main.PkgTests.Utils.temp_pkg_dir), ::Main.PkgTests.APITests.var"##60#70") at /home/sabae/src/Jotunheimr/Pkg.jl/test/utils.jl:34                                       
 [5] temp_pkg_dir(::Function) at /home/sabae/src/Jotunheimr/Pkg.jl/test/utils.jl:12                                                                                                                         
 [6] top-level scope at /home/sabae/src/Jotunheimr/Pkg.jl/test/api.jl:241                                                                                                                                   
 [7] top-level scope at /home/sabae/tmp/julia-build/julia-release-1.3/usr/share/julia/stdlib/v1.3/Test/src/Test.jl:1107                                                                                     
 [8] top-level scope at /home/sabae/src/Jotunheimr/Pkg.jl/test/api.jl:191  
```

I tracked this down far enough to finding that `_resolve()` is giving different answers for the exact same input set of packages.  I have no idea how this diff could cause that.

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
Great, so it passes remotely on Linux, fails locally on linux, and fails remotely on windows.

**Comment by 00vareladavid** [comment]
This seems related to the breakage in test isolation

**Comment by 00vareladavid** [comment]
Things were failing for me locally without #1334 and #1389 seems to fix the other case I was seeing. It might help to include those patches and retry

**Comment by staticfloat** [comment]
I don't think this is related to isolation; the resolver is giving a different result, choosing different package versions.

**Comment by staticfloat** [comment]
Confirmed that cherry-picking those two examples doesn't change anything; the resolver still spits out a different result than expected.

**Comment by staticfloat** [comment]
And testing on `arctic3` works.  Shrugging and moving on to the windows failures, which seem more legitimate.

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
bors try-

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
bors try

**Comment by staticfloat** [comment]
bors try-

**Comment by staticfloat** [comment]
bors try

File: src/PlatformEngines.jl [hunk 2]
```diff
@@ -849,7 +849,16 @@ function download_verify_unpack(url::AbstractString,
...
 
+        # Work around windows limitations regarding tempname()
         tarball_path = "$(tempname())-download.$(ext)"
+        tries = 0
+        while isfile(tarball_path) && tries < 100
+            tarball_path = "$(tempname())-download.$(ext)"
+            tries += 1
+        end
+        if tries >= 100
+            error("Unable to generate unused tempname! Clean up your temporary folder $(dirname(tempname())) and try again.")
```
Comment (KristofferC) [inline src/PlatformEngines.jl]: 🤣

**Comment by staticfloat** [comment]
Awesome. The tempname workaround fixed windows, and Mac failure is just DNS problems. I will drop the unrelated commits, squash and merge.

**Pull request merged**